src/reactA11yInputElementsRule.ts

```diff
@@ -47,30 +47,30 @@ export class Rule extends Lint.Rules.AbstractRule {
 }
 
 function isExcludedInputType(node: ts.JsxSelfClosingElement): boolean {
-    let isExcludedType = false;
-
     for (const attribute of node.attributes.properties) {
         if (tsutils.isJsxAttribute(attribute)) {
+            const isInputAttributeType = getJsxAttributesFromJsxElement(node).type;
             if (attribute.initializer !== undefined && tsutils.isStringLiteral(attribute.initializer)) {
                 const attributeText = attribute.initializer.text;
-                if (EXCLUDED_INPUT_TYPES.indexOf(attributeText) !== -1) {
-                    isExcludedType = true;
-                    return isExcludedType;
+                if (isInputAttributeType !== undefined && EXCLUDED_INPUT_TYPES.indexOf(attributeText) !== -1) {
+                    return true;
                 }
             }
         }
     }
-    return isExcludedType;
+    return false;
 }
 
 function walk(ctx: Lint.WalkContext<void>) {
     function cb(node: ts.Node): void {
         if (tsutils.isJsxSelfClosingElement(node)) {
             const tagName = node.tagName.getText();
 
-            if (tagName === 'input' && !isExcludedInputType(node)) {
+            if (tagName === 'input') {
                 const attributes = getJsxAttributesFromJsxElement(node);
-                if (isEmpty(attributes.value) && isEmpty(attributes.placeholder)) {
+                const isExcludedInputTypeValueEmpty = isEmpty(attributes.value) && isExcludedInputType(node);
+                const isPlaceholderEmpty = isEmpty(attributes.placeholder) && !isExcludedInputType(node);
+                if ((isEmpty(attributes.value) && isPlaceholderEmpty) || isExcludedInputTypeValueEmpty) {
                     ctx.addFailureAt(node.getStart(), node.getWidth(), MISSING_PLACEHOLDER_INPUT_FAILURE_STRING);
```

Comment:
😬 might this be simplified?

Reply:
A couple things to try:

* Pass `attributes` into `isExcludedInputType.type` so you don't have to call `getJsxAttributesFromJsxElement` each time
* Only make one call to `isExcludedInputType(node)` and store that in a variable, so you don't need to call it twice.

...but the test cases look good, build passes, and this is still a pretty simple rule; I'm fine with merging this in as is & filing a followup issue to look into cleaning it up. 

I'll wait for direction from you if you want to tackle it now - totally up to you! 😊

Reply:
Good call! Dried it up ⬇️